cmd/export.go

```diff
@@ -73,42 +84,131 @@ func exportCmd() *cobra.Command {
+	return fmt.Sprintf("output encoding [%s: [%s], %s: [%s]]",
+		formats.SPDXFORMAT, strings.Join(encodingOptions()[formats.SPDXFORMAT], ", "),
+		formats.CDXFORMAT, strings.Join(encodingOptions()[formats.CDXFORMAT], ", "),
+	)
+}
+
+func encodingOptions() map[string][]string {
+	return map[string][]string{
+		formats.CDXFORMAT:  {formats.JSON, formats.XML},
+		formats.SPDXFORMAT: {formats.JSON},
+	}
+}
+
+func formatHelp() string {
+	return fmt.Sprintf("output format [%s]", strings.Join(formatOptions(), ", "))
+}
+
+func formatOptions() []string {
+	spdxFormats := []string{
+		formats.SPDXFORMAT,
+		formats.SPDXFORMAT + "-2.3",
+	}
+
+	cdxFormats := []string{
+		formats.CDXFORMAT,
+		formats.CDXFORMAT + "-1.0",
+		formats.CDXFORMAT + "-1.1",
+		formats.CDXFORMAT + "-1.2",
+		formats.CDXFORMAT + "-1.3",
+		formats.CDXFORMAT + "-1.4",
+		formats.CDXFORMAT + "-1.5",
+	}
+
+	return append(spdxFormats, cdxFormats...)
+}
+
+func parseFormat(fs, encoding string) (formats.Format, error) {
+	results := map[string]string{}
+	pattern := regexp.MustCompile("^(?P<name>[^-]+)(?:-(?P<version>.*))?")
+	match := pattern.FindStringSubmatch(fs)
+
+	for idx, name := range match {
+		results[pattern.SubexpNames()[idx]] = name
+	}
+
+	baseFormat := results["name"]
+	version := results["version"]
+
+	if err := validateFormat(baseFormat); err != nil {
+		return formats.EmptyFormat, err
+	}
+
+	if err := validateEncoding(fs, encoding); err != nil {
+		return formats.EmptyFormat, err
+	}
+
+	var serializer native.Serializer
+
+	switch baseFormat {
+	case formats.CDXFORMAT:
+		if version == "" {
+			version = "1.5"
 		}
+
+		baseFormat = "application/vnd.cyclonedx"
+		serializer = serializers.NewCDX(version, encoding)
+	case formats.SPDXFORMAT:
+		if version == "" {
+			version = "2.3"
+		}
+
+		baseFormat = "text/spdx"
+		serializer = serializers.NewSPDX23()
 	}
+
+	format := formats.Format(fmt.Sprintf("%s+%s;version=%s", baseFormat, encoding, version))
+	writer.RegisterSerializer(format, serializer)
+
+	return format, nil
 }
 
-func initOpts(opts *export.Options, cfgFile, formatString, formatEncoding string) {
-	opts.CacheDir = viper.GetString("cache_dir")
-	opts.ConfigFile = cfgFile
-	opts.FormatString = formatString
-	opts.Encoding = formatEncoding
+func preRun(opts *options.Options) func(*cobra.Command, []string) {
+	return func(cmd *cobra.Command, _ []string) {
+		cfgFile, err := cmd.Flags().GetString("config")
+		cobra.CheckErr(err)
+
+		verbosity, err := cmd.Flags().GetCount("verbose")
+		cobra.CheckErr(err)
+
+		opts.
+			WithCacheDir(viper.GetString("cache_dir")).
+			WithConfigFile(cfgFile).
+			WithDebug(verbosity >= minDebugLevel)
+	}
 }
 
-func initBackend(opts *export.Options) *db.Backend {
-	backend := db.NewBackend(func(b *db.Backend) {
-		b.Options.DatabaseFile = filepath.Join(opts.CacheDir, db.DatabaseFile)
-		b.Options.Debug = opts.Debug
-		b.Logger = utils.NewLogger("export")
-	})
+func validateEncoding(fs, encoding string) error {
```

Comment:
```suggestion
func validateEncoding(fs string, encoding string) error {
```

Reply:
This causes `golangci-lint` to fail with

```
paramTypeCombine: func(fs string, encoding string) (formats.Format, error) could be replaced with func(fs, encoding string) (formats.Format, error) (gocritic)
```

we can talk about adjusting the linter config if there is consensus around it